sycl/include/sycl/detail/ur.hpp

```diff
@@ -0,0 +1,315 @@
+  /// a null-terminated string; target- and compiler-specific options
+  /// which are suggested to use to "compile" program at runtime
+  const char *CompileOptions;
+  /// a null-terminated string; target- and compiler-specific options
+  /// which are suggested to use to "link" program at runtime
+  const char *LinkOptions;
+  /// Pointer to the manifest data start
+  const char *ManifestStart;
+  /// Pointer to the manifest data end
+  const char *ManifestEnd;
+  /// Pointer to the target code start
+  const unsigned char *BinaryStart;
+  /// Pointer to the target code end
+  const unsigned char *BinaryEnd;
+  /// the offload entry table
+  _ur_offload_entry EntriesBegin;
+  _ur_offload_entry EntriesEnd;
+  // Array of preperty sets; e.g. specialization constants symbol-int ID map is
+  // propagated to runtime with this mechanism.
+  ur_device_binary_property_set PropertySetsBegin;
+  ur_device_binary_property_set PropertySetsEnd;
+  // TODO Other fields like entries, link options can be propagated using
+  // the property set infrastructure. This will improve binary compatibility and
+  // add flexibility.
+};
+using ur_device_binary = ur_device_binary_struct *;
+
+// Offload binaries descriptor version supported by this library.
+static const uint16_t UR_DEVICE_BINARIES_VERSION = 1;
+
+/// This struct is a record of all the device code that may be offloaded.
+/// It must match the __tgt_bin_desc structure generated by
+/// the clang-offload-wrapper tool when their Version field match.
+struct ur_device_binaries_struct {
+  /// version of this structure - for backward compatibility;
+  /// all modifications which change order/type/offsets of existing fields
+  /// should increment the version.
+  uint16_t Version;
+  /// Number of device binaries in this descriptor
+  uint16_t NumDeviceBinaries;
+  /// Device binaries data
+  ur_device_binary DeviceBinaries;
+  /// the offload entry table (not used, for compatibility with OpenMP)
+  _ur_offload_entry *HostEntriesBegin;
+  _ur_offload_entry *HostEntriesEnd;
+};
+using ur_device_binaries = ur_device_binaries_struct *;
+
+#ifdef XPTI_ENABLE_INSTRUMENTATION
+// Forward declarations
+namespace xpti {
+struct trace_event_data_t;
+}
+#endif
+
+namespace sycl {
+inline namespace _V1 {
+
+class context;
+
+namespace detail {
+
+namespace pi {
+// This function is deprecated and it should be removed in the next release
+// cycle (along with the definition for pi_context_extended_deleter).
+__SYCL_EXPORT void contextSetExtendedDeleter(const sycl::context &constext,
+                                             pi_context_extended_deleter func,
+                                             void *user_data);
+}
+
+class plugin;
+using PluginPtr = std::shared_ptr<plugin>;
+
+template <sycl::backend BE>
+__SYCL_EXPORT void *getPluginOpaqueData(void *opaquedata_arg);
+
+namespace ur {
+// Function to load a shared library
+// Implementation is OS dependent
+void *loadOsLibrary(const std::string &Library);
```

Comment:
We don't call `pi::` version of this function anywhere in SYCL headers, I think that this should be moved into our `os_util` part of the library. This could be done as a separate PR and then it won't be a part of PI -> UR transition at all.

Same applies to `unloadOsLibrary` and maybe other functions declared here.

Reply:
This was added as one of the follow up tasks mentioned in this [issue](https://github.com/intel/llvm/issues/14598).

Reply:
I also have #14791 to address this